File: java/src/main/java/sk/epostak/sdk/UblValidationException.java
package sk.epostak.sdk;

/**
 * Thrown when the API returns {@code 422} with
 * {@code error.code = "UBL_VALIDATION_ERROR"}.
 * <p>
 * The {@link #getRule()} method returns the first failing Peppol BIS 3.0 or
 * EN 16931 rule identifier. The human-readable error is in
 * {@link EPostakException#getMessage()}.
 *
 * <pre>{@code
 * try {
 *     client.documents().send(request);
 * } catch (UblValidationException e) {
 *     System.err.println("UBL rule violation: " + e.getRule());
 *     System.err.println("Message: " + e.getMessage());
 *     if (e.getRequestId() != null) {
 *         System.err.println("Request-Id: " + e.getRequestId());
 *     }
 * } catch (EPostakException e) {
 *     // other API errors
 * }
 * }</pre>
 */
public class UblValidationException extends EPostakException {

    /**
     * The rule identifier from the first validation failure, or {@code null}
     * if the response did not include one.
     */
    private final String rule;

    /**
     * Creates a new exception with all fields.
     *
     * @param status        HTTP status code (usually 422)
     * @param message       human-readable error message
     * @param code          machine-readable error code ({@code "UBL_VALIDATION_ERROR"})
     * @param details       additional structured error details, or {@code null}
     * @param type          RFC 7807 type, or {@code null}
     * @param title         RFC 7807 title, or {@code null}
     * @param detail        RFC 7807 detail, or {@code null}
     * @param instance      RFC 7807 instance, or {@code null}
     * @param requestId     server-assigned request id, or {@code null}
     * @param requiredScope OAuth scope required, or {@code null}
     * @param rule          first failing rule identifier (e.g. {@code "BR_02"}), or {@code null}
     */
    public UblValidationException(
            int status,
            String message,
            String code,
            Object details,
            String type,
            String title,
            String detail,
            String instance,
            String requestId,
            String requiredScope,
            String rule
    ) {
        super(status, message, code, details, type, title, detail, instance, requestId, requiredScope);
        this.rule = rule;
    }

    /**
     * Creates a new exception with status, message, requestId, and rule.
     * Convenience constructor for the common case.
     *
     * @param status    HTTP status code
     * @param message   human-readable error message
     * @param requestId server-assigned request id, or {@code null}
     * @param rule      first failing rule identifier, or {@code null}
     */
    public UblValidationException(int status, String message, String requestId, String rule) {
        this(status, message, "UBL_VALIDATION_ERROR", null, null, null, null, null, requestId, null, rule);
    }

    /**
     * Returns the first failing Peppol BIS 3.0 / EN 16931 rule identifier,
     * e.g. {@code "BR_02"}, {@code "PEPPOL_R008"}. May be {@code null} if the
     * server response did not include a rule field.
     *
     * @return the rule identifier, or {@code null}
     * @see UblRule
     */
    public String getRule() {
        return rule;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("UblValidationException{status=");
        sb.append(getStatus());
        if (rule != null) sb.append(", rule='").append(rule).append('\'');
        if (getRequestId() != null) sb.append(", requestId='").append(getRequestId()).append('\'');
        sb.append(", message='").append(getMessage()).append('\'');
        sb.append('}');
        return sb.toString();
    }
}

File: dotnet/src/EPostak/UblValidationException.cs
namespace EPostak;

/// <summary>
/// Well-known UBL pre-flight rule codes returned in <c>error.rule</c> when the
/// API rejects a document with <c>error.code == "UBL_VALIDATION_ERROR"</c>
/// (HTTP 422). Source of truth: <c>lib/ubl/generate.ts</c> in the epostak repo.
/// New rules may be added in future API versions; treat as hint, not closed enum.
/// </summary>
public static class UblRule
{
    /// <summary>BT-2 — Invoice issue date is mandatory.</summary>
    public const string BR_02 = "BR-02";

    /// <summary>BT-27 — Seller name is mandatory.</summary>
    public const string BR_05 = "BR-05";

    /// <summary>BT-44 — Buyer name is mandatory. Pass <c>receiverName</c> in the request body.</summary>
    public const string BR_06 = "BR-06";

    /// <summary>BT-31 / BT-32 — Seller VAT identifier required for VAT-rated invoices.</summary>
    public const string BR_11 = "BR-11";

    /// <summary>Invoice must have at least one line.</summary>
    public const string BR_16 = "BR-16";

    /// <summary>BT-1 — Invoice number must not be empty.</summary>
    public const string BT_1 = "BT-1";

    /// <summary>EndpointID empty — firm must have DIČ, IČO, or a registered Peppol ID.</summary>
    public const string PEPPOL_R008 = "PEPPOL-R008";
}

/// <summary>
/// Thrown when the API returns HTTP 422 with <c>code == "UBL_VALIDATION_ERROR"</c>
/// (or another <see cref="UblRule"/> code). Contains the rule identifier that
/// triggered the rejection and the optional server-assigned request ID.
/// </summary>
/// <remarks>
/// Catch this exception to handle schematron validation failures:
/// <list type="bullet">
/// <item><description>The offending rule identifier is exposed as <see cref="Rule"/>.</description></item>
/// <item><description>Human-readable details about each violation are in <see cref="EPostakException.Details"/>.</description></item>
/// <item><description>The server request ID (for support tickets) is in <see cref="EPostakException.RequestId"/>.</description></item>
/// </list>
/// </remarks>
/// <example>
/// <code>
/// try
/// {
///     await client.Documents.SendAsync(request);
/// }
/// catch (UblValidationException ex)
/// {
///     Console.WriteLine($"UBL rule violated: {ex.Rule}");
///     Console.WriteLine($"Details: {ex.Details}");
///     if (ex.RequestId is not null)
///         Console.WriteLine($"Support reference: {ex.RequestId}");
/// }
/// </code>
/// </example>
public sealed class UblValidationException : EPostakException
{
    /// <summary>
    /// The Peppol BIS / EN 16931 rule code that caused the rejection. One of
    /// the <see cref="UblRule"/> constants (e.g. <c>UblRule.BR_06</c>). May be
    /// an empty string if the server omitted the <c>rule</c> field.
    /// </summary>
    public string Rule { get; }

    internal UblValidationException(
        int status,
        string message,
        string rule,
        object? details,
        string? type,
        string? title,
        string? detail,
        string? instance,
        string? requestId,
        string? requiredScope)
        // Code stays "UBL_VALIDATION_ERROR" (always — that's the wire code);
        // the violated rule is exposed separately via Rule.
        : base(status, message, "UBL_VALIDATION_ERROR", details, type, title, detail, instance, requestId, requiredScope)
    {
        Rule = rule;
    }
}
